### ISSUE TITLE
ErrorHandler.php: Private members/methods and exception logging

### ISSUE THREAD

**Issue by icanhazpython** [Seldaek/monolog#858]
So maybe I missed a memo somewhere, but in my adventures in attempting to use the ErrorHandler class, I've run across the following issues:

- Members & at least one method is private, keeping me from easily extending it (and fixing the below without creating my own complete class). I get that inheritance is probably not the best way to fix things, LOL. But I would venture a guess that this class was never meant to be extended, and that's ok.
- More importantly, logging exceptions seems to not be handled properly. The class formats the exception message and stack trace by itself, but exception formatting with nicer stack trace formatting has already been implemented in NormalizerFormatter and descendants (specifically looking at you, LineFormatter). When passed a throwable object, the LineFormatter nicely formats the exception and stack trace into a multi-line format. By bypassing this behavior and formatting the stack trace by itself, ErrorHandler makes the stack trace look, well, not so nice. The call can be seen in ErrorHandler.php (5ce1c92) on line 153 in function 'handleException'. The exception object is passed as a context item, where it should be passed as a replacement for the formatted string. 

Thanks

**Comment by Seldaek** [comment]
I actually don't see what you mean here.. https://github.com/Seldaek/monolog/blob/1.x/src/Monolog/ErrorHandler.php#L130 uses the exception message/line as the log message, but then passes the exception as context info so that it should trigger nice formatting in the LineFormatter if you have one in use down in your handler stack. As far as I can tell this works fine?!

As for extending, well it's better to trigger discussions like these that can result in improvements IMO than having every user monkey patching stuff themselves :)

**Comment by icanhazpython** [comment]
You are correct sir, I must have done something incorrectly and saw the stack trace basically stringified in the context and not printed out in multiline form. When I did a quick 5-minute glance into the source, I incorrectly assumed that the exception normalizing in NormalizerFormatter was normalizing exceptions only on the $message/$data argument. I missed earlier that the $data argument contains the message, context, everything. I also missed that $message was being casted to string in the Logger class, so my stack traces were actually being converted to string by php's Exception class. Lol.. long day, sorry for the confusion. 

As to extending things, what I wanted to do was simply edit out certain sensitive data from logs. For example, if a PDO connection fails, the database credentials are printed in the first line of the stack trace. So I ended up doing something like this (catching the exception with my own set_exception_handler function, checking to see if it's a PDO connection issue, and choosing to format the stack trace myself or pass the exception onto monolog to do it) :

```php
if ($e instanceof PDOException && $e->getCode() == 2002) {
    $logger->error("Connection to database failed\n\nStack Trace:\n" . preg_replace('/PDO->__construct\(.+?\)/', 'PDO->__construct(*** CREDENTIALS REMOVED FOR SECURITY PURPOSES ***)', $e->getTraceAsString()) . "\n\n");
} else {
    $logger->error($e);
}
```

I suppose I could extend NormalizerFormatter and account for this particular case where it is normalizing the exceptions. Would be nice to have some kind of easy to use method to hide things such as this though. Any suggestions?
